external/fv3fit/tests/reservoir/test_domain.py

```diff
@@ -129,17 +149,89 @@ def test_stack_time_series_samples():
     np.testing.assert_array_equal(stacked[-1], np.array([9, 9, 9, 9]))
 
 
-def test_RankDivider_unstack_subdomain():
+@pytest.mark.parametrize(
+    "rank_extent, overlap, with_overlap",
+    [
+        ([5, 6, 6, 1], 1, True),
+        ([5, 6, 6, 2], 1, True),
+        ([5, 6, 6, 2], 1, False),
+        ([5, 6, 6, 2], 0, True),
+        ([1, 4, 4, 3], 0, False),
+    ],
+)
+def test_RankDivider_unstack_subdomain(rank_extent, overlap, with_overlap):
+    divider = RankDivider(
+        subdomain_layout=(2, 2),
+        rank_dims=["time", "x", "y", "z"],
+        rank_extent=rank_extent,
+        overlap=overlap,
+    )
+    rank_arr = np.arange(np.prod(rank_extent)).reshape(*rank_extent)
+    subdomain_arr = divider.get_subdomain_tensor_slice(
+        rank_arr, 0, with_overlap=with_overlap
+    )
+    stacked = stack_time_series_samples(subdomain_arr)
+    assert len(stacked.shape) == 2
+    np.testing.assert_array_equal(
+        divider.unstack_subdomain(stacked, with_overlap=with_overlap), subdomain_arr
+    )
+
+
+def test_RankDivider_flatten_subdomains_to_columns():
     nt, nx, ny, nz = 5, 6, 6, 2
     divider = RankDivider(
         subdomain_layout=(2, 2),
         rank_dims=["time", "x", "y", "z"],
         rank_extent=[nt, nx, ny, nz],
         overlap=1,
     )
-    rank_arr = np.random.randn(nt, nx, ny, nz)
-    subdomain_arr = divider.get_subdomain_tensor_slice(rank_arr, 0, with_overlap=True)
-    stacked = stack_time_series_samples(subdomain_arr)
-    np.testing.assert_array_equal(
-        divider.unstack_subdomain(stacked, with_overlap=True), subdomain_arr
+    input = np.array(
+        [[0, 1, 10, 11], [2, 3, 12, 13], [20, 21, 30, 31], [22, 23, 32, 33]]
+    )
+    input_with_halo = np.pad(input, pad_width=1)
+    # add time dim of length 2
+    input_with_halo = np.stack([input_with_halo, input_with_halo], axis=0)
+    flattened = divider.flatten_subdomains_to_columns(
+        input_with_halo, with_overlap=True
+    )
+
+    # 4 subdomains each with x, y dims (4, 4)
+    assert flattened.shape == (2, 16, 4)
+    # Check first subdomain of first sample has correct values
+    np.testing.assert_array_almost_equal(
+        flattened[0, :, 0],
+        np.array([0, 0, 0, 0, 0, 0, 1, 10, 0, 2, 3, 12, 0, 20, 21, 30]),
+    )
```

Comment:
I know it's tedious, but it feels like you might as well test that all subdomains have the values you expect.